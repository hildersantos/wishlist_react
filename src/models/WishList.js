import { types } from "mobx-state-tree";

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: ""
  })
  .actions(self => ({
    changeName(newName) {
      return (self.name = newName);
    },
    changePrice(newPrice) {
      return (self.price = newPrice);
    },
    changeImage(newImage) {
      return (self.image = newImage);
    }
  }));

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  })
  .actions(self => ({
    addItem(item) {
      self.items.push(item);
    }
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((sum, entry) => sum + entry.price, 0);
    }
  }));
