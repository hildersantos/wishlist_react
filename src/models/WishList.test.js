import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { WishListItem, WishList } from "./WishList";

const data = {
  name: "Cronicas de Narnia",
  price: 39.99
};
it("can create an instance of a model", () => {
  const item = WishListItem.create(data);

  expect(item.price).toBe(39.99);
  expect(item.image).toBe("");
});

it("can create an item", () => {
  const list = WishList.create({});
  const states = [];
  onSnapshot(list, snapshot => {
    states.push(snapshot);
  });

  list.addItem({
    name: "O Alquimista",
    price: 21.0
  });

  expect(list.items.length).toBe(1);
  expect(list.items[0].name).toBe("O Alquimista");
  list.items[0].changeName("Diario de um Mago");

  expect(getSnapshot(list)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
});

it("can create an item - 2", () => {
  const list = WishList.create({});
  const patches = [];
  onPatch(list, patch => {
    patches.push(patch);
  });

  list.addItem({
    name: "O Alquimista",
    price: 21.0
  });

  list.items[0].changeName("Diario de um Mago");

  expect(patches).toMatchSnapshot();
});

it("can change item data", () => {
  const item = WishListItem.create(data);

  item.changeName("O Hobbit");
  item.changePrice(40.0);
  item.changeImage("");

  expect(item.name).toBe("O Hobbit");
  expect(item.price).toBe(40.0);
  expect(item.image).toBe("");
});

it("can create an wishlist", () => {
  const wishlist = WishList.create({
    items: [data]
  });

  expect(wishlist.items.length).toBe(1);
  expect(wishlist.items[0].price).toBe(39.99);
});
