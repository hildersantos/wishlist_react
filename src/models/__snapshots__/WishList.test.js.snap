// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create an item - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "O Alquimista",
      "price": 21,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Diario de um Mago",
  },
]
`;

exports[`can create an item 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Diario de um Mago",
      "price": 21,
    },
  ],
}
`;

exports[`can create an item 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "O Alquimista",
        "price": 21,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Diario de um Mago",
        "price": 21,
      },
    ],
  },
]
`;
